### PR TITLE
ENH: Add xlsb auto detection to read_excel and respect default options

### LINKED ISSUE #34252: pd.set_option("io.excel.xlsx.reader", "openpyxl") does not work
This works fine for an xlsx file:

`pd.read_excel(path, engine="openpyxl")`

However if i try using set_option then read_excel fails:
```
pd.set_option("io.excel.xlsx.reader", "openpyxl")
pd.read_excel(path)
```
ImportError: Missing optional dependency 'xlrd'. Install xlrd >= 1.0.0 for Excel support Use pip or conda to install xlrd.

### LINKED ISSUE #35416: ENH: Auto Detect engine in read_excel
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

This works:

```python
import pandas as pd
df = pd.read_excel('Book2.xlsb', engine='pyxlsb')
```

This doesn't:

```python
import pandas as pd
df = pd.read_excel('Book2.xlsb')
```

#### Problem description

All supported Excel formats (`xls`, `xlsx`, `xlsm`) have a default engine based on the extension, so you can simply do:

```
df = pd.read_excel('Book2.xlsx')
```

For `xlsb`,  when you do `df = pd.read_excel('Book2.xlsb')`, you get this error:

```
XLRDError: Excel 2007 xlsb file; not supported
```

#### Expected Output

No error, i.e. it should figure out that for `xlsb` extensions, `pyxlsb` is the default engine.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None

pandas           : 1.0.5
numpy            : 1.18.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 49.2.0.post20200714
Cython           : None
pytest           : 5.4.3
hypothesis       : None
sphinx           : 2.4.0
blosc            : None
feather          : None
xlsxwriter       : 1.2.7
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.16.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.3
numexpr          : None
odfpy            : None
openpyxl         : 3.0.4
pandas_gbq       : None
pyarrow          : 0.17.1
pytables         : None
pytest           : 5.4.3
pyxlsb           : 1.0.6
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
tabulate         : 0.8.7
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.2.7
numba            : None

</details>

### LINKED ISSUE #38667: BUG: read_excel does not use pyxlsb for xlsb files when engine is None
From `pandas.tests.io.excel.test_excel_read_binary_via_read_excel` (currently only in #38571):

```
        if read_ext == ".xlsb" and engine == "pyxlsb":
            with pytest.raises(IOError, match="File contains no valid workbook part"):
                with open("test1" + read_ext, "rb") as f:
                    pd.read_excel(f)
```

This should instead determine that pyxlsb should be used when installed and no error should be raised.

### PR DESCRIPTION
**Pull request by lithomas1**
- [x] closes #35416 , closes #34252
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Note: Even though most of the reader options for read_excel only have 1 option, I think it is still useful to respect pandas.set_option. 

Also, if some people have older xlrd(e.g. 1.2.0) and don't have openpyxl, they can use options to force pandas to read xlsx files with xlrd. (this is possible now, but xlrd's maintainer has made it clear he doesn't support for things like this)

### CODE REVIEW

**Comment by lithomas1** [comment]
cc @WillAyd @TomAugspurger

File: doc/source/whatsnew/v1.3.0.rst [hunk 1]
```diff
@@ -45,6 +45,8 @@ Other enhancements
 - Improved consistency of error message when passing an invalid ``win_type`` argument in :class:`Window` (:issue:`15969`)
 - :func:`pandas.read_sql_query` now accepts a ``dtype`` argument to cast the columnar data from the SQL database based on user input (:issue:`10285`)
 - Improved integer type mapping from pandas to SQLAlchemy when using :meth:`DataFrame.to_sql` (:issue:`35076`)
+- :func:`pandas.read_excel` can now auto detect .xlsb files (:issue:`35416`)
+- :func:`pandas.read_excel` now respects ``pandas.set_option`` (:issue:`34242`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:49]: not sure what this is in reference, can you clarify? (that is an unrelated issue).

File: pandas/io/excel/_base.py [hunk 2]
```diff
@@ -1003,6 +1003,8 @@ class ExcelFile:
              then `odf <https://pypi.org/project/odfpy/>`_ will be used.
            - Otherwise if ``path_or_buffer`` is an xls format,
              ``xlrd`` will be used.
+           - Otherwise if ``path_or_buffer`` is in xlsb format,
```
Comment (jreback) [inline pandas/io/excel/_base.py:1006]: add a versionadded 1.3

File: pandas/io/excel/_util.py [hunk 3]
```diff
@@ -23,32 +23,38 @@ def register_writer(klass):
...
+        Whether to get the default engine for reading or writing.
+        Either 'read' or 'write'
 
     Returns
     -------
     str
         The default engine for the extension.
     """
-    _default_writers = {
+    _default_engines = {
```
Comment (jreback) [inline pandas/io/excel/_util.py:43]: can we just make a separate reader/write dict here to make this really clear

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/excel/_base.py [hunk 4]
```diff
@@ -1064,21 +1066,12 @@ def __init__(
...
-                    import_optional_dependency(
-                        "openpyxl", raise_on_missing=False, on_version="ignore"
-                    )
-                    is not None
-                ):
-                    engine = "openpyxl"
-                else:
-                    engine = "xlrd"
+            try:
+                engine = config.get_option(f"io.excel.{ext}.reader", silent=True)
```
Comment (rhshadrach) [inline pandas/io/excel/_base.py:1070]: What is the error message if this is set to a nonexistent engine?

File: pandas/io/excel/_base.py [hunk 5]
```diff
@@ -1106,7 +1099,6 @@ def __init__(
                     FutureWarning,
                     stacklevel=stacklevel,
                 )
-        assert engine in self._engines, f"Engine {engine} not recognized"
```
Comment (rhshadrach) [inline pandas/io/excel/_base.py:1109]: Why remove?

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Overall looks good. The behavior of respecting `io.excel.{ext}.reader` needs tests, including a test for the error message when it is set to a nonexistent engine.

File: doc/source/whatsnew/v1.3.0.rst (hunk 1 above)
Comment (rhshadrach) [reply doc/source/whatsnew/v1.3.0.rst:49]: I've closed #38667 as a duplicate of #35416, the OP should be updated to link to the latter instead.

File: pandas/io/excel/_base.py (hunk 4 above)
Comment (lithomas1) [reply pandas/io/excel/_base.py:1070]: I thought this would raise, but it doesn't apparently since pd.set_option doesn't validate options for some reason. I will add such validation into this PR.

File: pandas/io/excel/_base.py (hunk 5 above)
Comment (lithomas1) [reply pandas/io/excel/_base.py:1109]: Engine should be validate by https://github.com/pandas-dev/pandas/pull/38710/files#diff-63200ddb7f5656b8ee868a28d9cb7720ffe50689b0e3fb0b4e15cc5c0ae80dd7R1035 if passed explicitly
and by set_option if set there(I will add this).

File: doc/source/whatsnew/v1.3.0.rst (hunk 1 above)
Comment (lithomas1) [reply doc/source/whatsnew/v1.3.0.rst:49]: @rhshadrach That issue should remain open until this is merged since the test still needs to be updated.

File: pandas/io/excel/_base.py (hunk 2 above)
Comment (lithomas1) [reply pandas/io/excel/_base.py:1006]: @jreback this is already part of a versionchanged block(1.2.0). Should i just update the versionchanged to 1.3.0?

Comment (jreback) [reply pandas/io/excel/_base.py:1006]: you can add a version updated 1.2 / alt i guess this is a backport so would be ok for 1.2.1

File: doc/source/whatsnew/v1.3.0.rst (hunk 1 above)
Comment (lithomas1) [reply doc/source/whatsnew/v1.3.0.rst:49]: @jreback, these are two somewhat related issues as they both pertain to engine=None. I figured I would solve both of them here. (Respecting set_option is a one-liner)

File: pandas/io/excel/_base.py (hunk 4 above)
Comment (lithomas1) [reply pandas/io/excel/_base.py:1070]: Fixed, it will be of the form "Value Error: Value must be one of" and the valid engines list.

File: doc/source/whatsnew/v1.3.0.rst (hunk 1 above)
Comment (rhshadrach) [reply doc/source/whatsnew/v1.3.0.rst:49]: #34242 refers to `NaT+Period doesnt match Period+NaT`, needs to be fixed. For #38667, it is a duplicate of #35416. #35416 will remain open until this is merged.

File: pandas/io/excel/_util.py [hunk 6]
```diff
@@ -23,32 +23,46 @@ def register_writer(klass):
...
+        Whether to get the default engine for reading or writing.
+        Either 'read' or 'write'
 
     Returns
     -------
     str
         The default engine for the extension.
     """
+    _default_readers = {
+        "xlsx": "openpyxl",
```
Comment (rhshadrach) [inline pandas/io/excel/_util.py:44]: Looks to me like this changes current behavior. On master, if xlrd is installed and version < 2.0 and openpyxl is not, will currently default to xlrd. In this PR, this will error instead, breaking user code.

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: pandas/tests/io/excel/test_readers.py [hunk 7]
```diff
@@ -1229,3 +1227,10 @@ def test_read_datetime_multiindex(self, engine, read_ext):
         expected = DataFrame([], columns=expected_column_index)
 
         tm.assert_frame_equal(expected, actual)
+
+    def test_engine_invalid_option(self, read_ext):
+        # read_ext includes the '.' hence the weird formatting
+        engine = pd.get_option(f"io.excel{read_ext}.reader")
```
Comment (jreback) [inline pandas/tests/io/excel/test_readers.py:1233]: can you instead use the context manager to test options. this is likely causing side effects.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.3.0.rst (hunk 1 above)
Comment (lithomas1) [reply doc/source/whatsnew/v1.3.0.rst:49]: Oh i meant #34252, I'll fix this.

File: pandas/io/excel/_util.py [hunk 8]
```diff
@@ -23,32 +23,46 @@ def register_writer(klass):
...
+    mode : str
+        Whether to get the default engine for reading or writing.
+        Either 'read' or 'write'
 
     Returns
     -------
     str
         The default engine for the extension.
     """
+    _default_readers = {
```
Comment (jreback) [inline pandas/io/excel/_util.py:43]: hmm maybe this could integrate with the `_writers` and `_readers` module level variables?

**Comment by lithomas1** [comment]
FYI @rhshadrach, I'm going to be changing the behavior in this PR if both xlrd and openpyxl aren't installed.
Instead of raising, ``ImportError: Missing optional dependency 'xlrd'. Install xlrd >= 1.0.0 for Excel support Use pip or conda to install xlrd.`` (which is wrong anyways, we require xlrd>=1.2.0, this will need to be fixed in another PR)
We now raise, ``ImportError: Missing optional dependency 'openpyxl'.  Use pip or conda to install openpyxl.``

This way, we encourage users to use openpyxl not xlrd.

File: pandas/io/excel/_util.py (hunk 6 above)
Comment (lithomas1) [reply pandas/io/excel/_util.py:44]: Fixed and tested it with xlrd==1.2.0 and no openpyxl.

File: pandas/io/excel/_util.py (hunk 8 above)
Comment (lithomas1) [reply pandas/io/excel/_util.py:43]: @jreback I'm not sure how these are related. The _writers dict at the top of this file is simply a mapping between engine names and there class names and I don't know where to find the _readers dict. Can you please elaborate on this more?

File: pandas/tests/io/excel/test_readers.py (hunk 7 above)
Comment (lithomas1) [reply pandas/tests/io/excel/test_readers.py:1233]: @jreback can you point me to some documentation on how to use this? I am currently following the approach used here. 
https://github.com/pandas-dev/pandas/pull/31569/files#diff-5974844fa02c9b81f3477198023bd02503bf82da1e3d15df6dfd62cdca48666bR242

File: doc/source/whatsnew/v1.3.0.rst [hunk 9]
```diff
@@ -51,6 +51,8 @@ Other enhancements
 - :func:`pandas.read_sql_query` now accepts a ``dtype`` argument to cast the columnar data from the SQL database based on user input (:issue:`10285`)
 - Improved integer type mapping from pandas to SQLAlchemy when using :meth:`DataFrame.to_sql` (:issue:`35076`)
 - :func:`to_numeric` now supports downcasting of nullable ``ExtensionDtype`` objects (:issue:`33013`)
+- :func:`pandas.read_excel` can now auto detect .xlsb files (:issue:`35416`)
+- :func:`pandas.read_excel` now respects ``pandas.set_option`` (:issue:`34252`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:55]: this should be in Bug Fixes I/O (2nd line)

File: pandas/io/excel/_base.py [hunk 10]
```diff
@@ -1064,21 +1072,18 @@ def __init__(
...
-                else:
-                    engine = "xlrd"
+            if (
+                import_optional_dependency(
+                    "openpyxl", raise_on_missing=False, on_version="warn"
+                )
+                is None
+                and xlrd_version is not None
+            ):
+                config.set_option("io.excel.xlsx.reader", "xlrd")
```
Comment (jreback) [inline pandas/io/excel/_base.py:1082]: hmm this is mutating a user option. can you rather just set engine here?

File: pandas/io/excel/_base.py [hunk 11]
```diff
@@ -1064,21 +1072,18 @@ def __init__(
...
+            # ext will always be valid, otherwise inspect_excel_format would raise
+            engine = config.get_option(f"io.excel.{ext}.reader", silent=True)
+            if engine == "auto":
+                engine = get_default_engine(ext, mode="read")
```
Comment (jreback) [inline pandas/io/excel/_base.py:1086]: read -> reader

File: pandas/io/excel/_util.py [hunk 12]
```diff
@@ -23,32 +23,46 @@ def register_writer(klass):
...
+def get_default_engine(ext, mode="read"):
     """
-    Return the default writer for the given extension.
+    Return the default reader/writer for the given extension.
 
     Parameters
     ----------
     ext : str
         The excel file extension for which to get the default engine.
+    mode : str
```
Comment (jreback) [inline pandas/io/excel/_util.py:34]: user reader / writer

File: pandas/io/excel/_util.py (hunk 8 above)
Comment (jreback) [inline pandas/io/excel/_util.py:43]: can you assert mode in ['reader' ,'writer']

File: pandas/io/excel/_util.py (hunk 12 above)
Comment (jreback) [reply pandas/io/excel/_util.py:34]: mode: str {'reader', 'writer'}

File: pandas/io/excel/_util.py (hunk 8 above)
Comment (jreback) [reply pandas/io/excel/_util.py:43]: no i mean the _default_readers/writers could simply be module level as well. but i guess ok here too.

File: pandas/io/excel/_util.py [hunk 13]
```diff
@@ -23,32 +23,46 @@ def register_writer(klass):
...
         "ods": "odf",
     }
-    xlsxwriter = import_optional_dependency(
-        "xlsxwriter", raise_on_missing=False, on_version="warn"
-    )
-    if xlsxwriter:
-        _default_writers["xlsx"] = "xlsxwriter"
-    return _default_writers[ext]
+    if mode == "write":
+        xlsxwriter = import_optional_dependency(
```
Comment (jreback) [inline pandas/io/excel/_util.py:58]: why is this checked here cc @rhshadrach this is just a mapping right? when the engine is constructed this should actually be checked.

Comment (jreback) [reply pandas/io/excel/_util.py:58]: IOW get_default_engine should never raise (or if we choose to do this, then it should be documented, and if this is the case). why are we not importing openpyxl for readers here? its just pretty inconsistent.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/io/excel/test_readers.py (hunk 7 above)
Comment (jreback) [reply pandas/tests/io/excel/test_readers.py:1233]: use pd.option_context

File: pandas/io/excel/_base.py (hunk 10 above)
Comment (rhshadrach) [reply pandas/io/excel/_base.py:1082]: I think this should be moved to `get_default_engine`. It should just adjust the return value, not modify the user option.

File: pandas/tests/io/excel/test_readers.py (hunk 7 above)
Comment (jreback) [reply pandas/tests/io/excel/test_readers.py:1233]: https://github.com/pandas-dev/pandas/issues/38813 that is not a good example to follow

File: pandas/io/excel/_util.py (hunk 13 above)
Comment (rhshadrach) [reply pandas/io/excel/_util.py:58]: This doesn't raise. Rather, it changes the default engine based on what the user has installed. So the logic for xlsx files is "if user has xlsxwriter installed, use it as the default; otherwise use openpyxl".

Comment (jreback) [reply pandas/io/excel/_util.py:58]: ahh ok makes sense, ok then

@lithomas1 can you add a commen to this effect here

Comment (lithomas1) [reply pandas/io/excel/_util.py:58]: Yeah, this is old code, from when the function was get_default_writers. I'll add in the comment.

File: pandas/io/excel/_util.py [hunk 14]
```diff
@@ -23,32 +23,61 @@ def register_writer(klass):
...
+            _default_writers["xlsx"] = "xlsxwriter"
+        return _default_writers[ext]
+    else:
+        if (
+            import_optional_dependency(
+                "openpyxl", raise_on_missing=False, on_version="warn"
+            )
+            is None
+            and import_optional_dependency(
+                "xlrd", raise_on_missing=False, on_version="raise"
```
Comment (rhshadrach) [inline pandas/io/excel/_util.py:73]: We shouldn't be raising here. If a user doesn't have openpyxl/xlrd installed, this will raise even if a user isn't reading an xlsx file. I don't think we should even be warning here (although I see that warnings are given for the writer code above); warnings/errors should only be raised if the outdated/not-installed package is attempted to be used.

Comment (rhshadrach) [reply pandas/io/excel/_util.py:73]: @jreback - thoughts?

File: doc/source/whatsnew/v1.3.0.rst [hunk 15]
```diff
@@ -266,7 +267,7 @@ I/O
 - Bug in :func:`to_hdf` raising ``KeyError`` when trying to apply
   for subclasses of ``DataFrame`` or ``Series`` (:issue:`33748`).
 - Bug in :func:`json_normalize` resulting in the first element of a generator object not being included in the returned ``DataFrame`` (:issue:`35923`)
-
+- :func:`pandas.read_excel` now respects ``pandas.set_option`` (:issue:`34252`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v1.3.0.rst:270]: Make ``:func:`pandas.set_option` `` so a link is made to the set_option page.

File: pandas/io/excel/_base.py [hunk 16]
```diff
@@ -123,6 +123,10 @@
          then `odf <https://pypi.org/project/odfpy/>`_ will be used.
        - Otherwise if ``path_or_buffer`` is an xls format,
          ``xlrd`` will be used.
+       - Otherwise if ``path_or_buffer`` is in xlsb format,
+         ``pyxlsb`` will be used.
```
Comment (rhshadrach) [inline pandas/io/excel/_base.py:127]: Let's link to `` `pyxlsb <https://pypi.org/project/pyxlsb/>`_ `` here, as is with the other engines.

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: pandas/io/excel/_util.py (hunk 14 above)
Comment (jreback) [reply pandas/io/excel/_util.py:73]: yeah this is solely to change the default.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/excel/_util.py [hunk 17]
```diff
@@ -23,32 +23,61 @@ def register_writer(klass):
...
+            )
+            is None
+            and import_optional_dependency(
+                "xlrd", raise_on_missing=False, on_version="ignore"
+            )
+            is not None
+        ):
+            # if no openpyxl but xlrd installed, return xlrd
+            # the version is handled elsewhere
+            _default_readers["xlsx"] = "xlrd"
```
Comment (jreback) [inline pandas/io/excel/_util.py:79]: this is not correct. xlrd should not be the default handler anymore of xlsx

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/excel/_util.py (hunk 17 above)
Comment (rhshadrach) [reply pandas/io/excel/_util.py:79]: The behavior here agrees with master: xlrd only gets set as the default if openpyxl is not installed, as not using xlrd may break user code. We then either warn or raise depending on the version of xlrd in _base.py.

Are we wanting to change so that xlrd is never the default for 1.3?

Comment (jreback) [reply pandas/io/excel/_util.py:79]: ok, then that's fine for this PR

>A re we wanting to change so that xlrd is never the default for 1.3?

if 1.3 == 2.0 then for sure (still TBD), but even if not I would be leaning that way. Why dont' you open an issue

Comment (rhshadrach) [reply pandas/io/excel/_util.py:79]: Opened #38884

**Review by rhshadrach — APPROVED** [review]
lgtm, thanks @lithomas1!

**Comment by jreback** [comment]
is there a test for this https://github.com/pandas-dev/pandas/issues/34252, e.g. that the set option works correctly? (i see once added for an invalid value), and am pretty sure we are testing this, but just want to confirm.

**Comment by lithomas1** [comment]
@jreback, 
I don't think there is a way to know which engine read_excel uses, so I didn't add a test.

**Comment by rhshadrach** [comment]
@lithomas1 `read_excel` is monkey patched in the tests for this purpose. Use `pd.read_excel.keywords["engine"]`.

**Comment by rhshadrach** [comment]
Ack - my previous comment was incorrect. That is of course only for which engine is passed to read_excel.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @lithomas1 very nice!
